crates/hyperqueue/src/dashboard/ui/widgets/chart.rs

```diff
@@ -0,0 +1,217 @@
+use chrono::{DateTime, Local};
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use tako::common::Map;
+use tui::layout::Rect;
+use tui::style::{Color, Modifier, Style};
+use tui::symbols::Marker;
+use tui::text::Span;
+use tui::widgets::{Axis, Block, Borders, Chart, Dataset};
+
+use crate::dashboard::data::DashboardData;
+use crate::dashboard::ui::styles::chart_style_deselected;
+use crate::dashboard::ui::terminal::DashboardFrame;
+
+/// Generic chart that shows graphs described by different `ChartPlotter`.
+pub struct DashboardChart {
+    /// The Chart's name
+    label: String,
+    /// The charts shown on the `DashboardChart`.
+    chart_plotters: Vec<Box<dyn ChartPlotter>>,
+    /// Chart's data, for each of the different plots on the chart.  
+    datasets: Map<String, Vec<(f64, f64)>>,
+    /// The style of a plot on the chart.
+    dataset_styles: Map<String, PlotStyle>,
+    /// The end time of the data displayed on the chart,
+    end_time: SystemTime,
+    /// The duration for which the data is plotted for.
+    view_size: Duration,
+    /// if true, the `end_time` is always updated to current time.
+    is_live: bool,
+}
+
+#[derive(Copy, Clone)]
+pub struct PlotStyle {
+    pub color: Color,
+    pub marker: Marker,
+}
+
+/// Declares a group of charts plotted on the `DashboardChart` with `get_charts`
+/// and defines how data will be fetched for each chart with `fetch_data_points`
+/// Multiple `ChartPlotters` can be used to add different charts to `DashboardChart`
+/// with each of the charts fetching their own data.
+pub trait ChartPlotter {
+    /// Declares the labels of the charts that will be plotted by this plotter and their style.
+    fn get_charts(&self) -> Map<String, PlotStyle>;
+
+    /// Fetches one set of points for each of the keys returned by `get_charts`
+    /// at given `time`
+    fn fetch_data_points(&self, data: &DashboardData, time: SystemTime) -> Map<String, (f64, f64)>;
```

Comment:
It seems to me that a more efficient API would be to get a range of timepoints for a given key rather than get all keys for a timepoint. But until it becomes a bottleneck we don't need to change it.